Grammar documentation (#21)
langium/langium-website: content/docs/grammar-language.md

```
person 25 Bob
```
will throw an error.

However, it is possible to declare a group of properties in an unordered fashion using the `&` operator
```
Person:
    'person' name=ID & age=INT
```
will now allow `name` and `age` to be declared in any order.
```
person 25 Bob
```
will then successfully create an object of type `Person`.

Cardinality (?,*,+ operators) also applies to unordered group. Please note that assignments with a cardinality of `+` or `*` have to appear continuously and cannot be interrupted by an other assignment and resumed later.

#### Simple Actions
It is possible for a rule to return different types depending on declaration
```
RuleOne returns TypeOne:
    'keywordOne' name=ID | RuleTwo;

RuleTwo returns TypeTwo:
    'keywordTwo' name=ID;
```
In the above example, we rely on a *rule call* to specify the return type. With more complex rules, the readability will be highly impacted. *Actions* allow to improve the readability of the grammar by explicitly defining the return type. *Actions* are declared inside of curly brackets `{}`:
```
RuleName returns TypeOne:
    'keywordOne' name=ID | 'keywordTwo' {TypeTwo} name=ID;
```

#### Unassigned Rule Calls

> Parser rules do not necessarily need to return an object, they can also refer to other parser rules which in turn will be responsible for returning the object.

Reviewer: Proposal: move this subsection before _Simple Actions_

```suggestion
Parser rules do not necessarily need to create an object, they can also refer to other parser rules which in turn will be responsible for returning the object.
```

```
The parser rule `AbstractDefinition` will not create an object of type AbstractDefinition. Instead, it calls either the `Definition` or `DeclaredParameter` parser rule which will be responsible for creating an object of a given type (or call other parser rules if they are unassigned rule calls themselves). 

#### Assigned Actions
The parser is built using [Chevrotain](https://github.com/chevrotain/chevrotain) which implements a LL(k) parsing algorithm (left-to-right). By definition, a LL(k) grammar cannot have rules containing left recursion.

Consider the following: 
```
Addition:
    Addition '+' Addition | '(' Addition ')' | value=INT;
```
The parser rule `Addition` is left-recursive and will not be parsable. We can go around this issue by *left-factoring* the rule, *i.e.* by factoring out the common left-factor. We introduce a new rule `Expression`:
```
Expression:
    '(' Addition ')' | value=INT;

Addition:
    Expression ('+' Expression)*;
```

[NOTE] I left out the part regarding `tree rewrite actions` as I think it could be part of a more advanced documentation part on the AST/CST. I can add it if we think it belongs here.

## Data Type Rules

> Data type rules are similar to terminal rules as they match a sequence of characters. However, they are parser rules and therefore are context-dependent, and are allowed to use hidden terminal rules. Contrary to terminal rules, they cannot use *regular expressions* to match a stream of character and have to compose with terminal rules.

Reviewer: ```suggestion
Data type rules are similar to terminal rules as they match a sequence of characters. However, they are parser rules and therefore are context-dependent, and are allowed to use hidden terminal rules. Contrary to terminal rules, they cannot use *regular expressions* to match a stream of character and have to be composed of terminal rules and keywords.
```

#### Character Range
The operator `..` is used to declare a character range. It is equivalent to the operator `-` in regular expression. It matches any character in between the left character and the right character (inclusive on both ends).
```
terminal INT returns number: ('0'..'9')+;
```
is equivalent to the regular expression:
```
terminal INT returns number: /[0-9]+/;
```
Here, `INT` is matched to one or more characters (by using the operand `+`, which defines a [cardinality](#cardinalities) of 'one or many') between `0` and `9` (inclusive on both ends).

#### Wildcard
The operator `.` is used to match any character and is similar in regular expression.
```
terminal HASHTAG: '#'.+;
```
In this example, the terminal rule `HASHTAG` matches a sequence of character starting with `#` followed by one or many ([cardinality](#cardinalities) +) characters.

Equivalent in regular expression:
```
terminal HASHTAG: /#.+/;
```
#### Until Token
The operator `->` indicates that all characters should be consumed from the left token *until* the right token occurs. For example, the terminal rule for multi-line comment can be implemented as:
```
terminal ML_COMMENT: '/*' -> '*/';
```

> There is no direct equivalent in regular expression for the until token.

Author: Added equivalence in commit [5487c8f](https://github.com/langium/langium-website/pull/21#:~:text=until%20token%20equivalence-,5487c8f,-Merge%20state)

#### Negated Token
It is possible to negate all tokens using the operator `!`, in regular expression the operator `^` is used for negation.
```
terminal BETWEEN_HASHES:'#' (!'#')* '#';
```
In regular expression:
```
terminal BETWEEN_HASHES: /#[^#]*#/;
```

#### Rule Calls
A terminal rule can implement other terminal rules. 
```
terminal DOUBLE returns number: INT '.' INT;
```
Note: it is easy to create conflict between terminal rules by using *terminal rule calls*. See [Data Type Rules](#data-type-rules) for further details.

#### Alternatives
It is possible to match several valid options. The terminal rule `STRING` can use alternatives to match a sequence of characters between double quotes `""` or single quotes `''`:
```
terminal STRING: /"[^"]*"|'[^']*'/;
```

#### Groups
Tokens can be put in sequence specifying the order they have to appear
```
terminal FLIGHT_NUMBER: ('A'..'Z')('A'..'Z')('0'..'9')('0'..'9')('0'..'9')('0'...'9')?;
```

> In this example, the token `FLIGHT_NUMBER` **must** start with two capital letters and followed by three of four digits.

Reviewer 2: Minor:
```suggestion
In this example, the token `FLIGHT_NUMBER` **must** start with two capital letters and followed by three or four digits.
```

### Terminal Fragments

> Fragments allow for sub-definition of terminal rules to be extracted. They are not consumed by the lexer and have to be consumed by other terminal rules.

Reviewer 2: Suggestion: Add here, that normal terminals can also be used as part of other terminals, not only fragments.

Author: Isn't that what is documented in the "rule calls" section (101-106) ? Or am I missing something?

```
hidden terminal WS: /\s+/;
hidden terminal ML_COMMENT: /\/\*[\s\S]*?\*\//;
hidden terminal SL_COMMENT: /\/\/[^\n\r]*/;
```

## Parser Rules
While [terminal rules](#terminal-rules) indicate to the lexer what sequence of characters are valid tokens, *parser rules* indicate to the parser what sequence of tokens are valid. Parser rules lay the structure of objects to be created by the parser and result in the creation of the *abstract syntax tree* which represents the syntactic structure of your language. In Langium, parser rules are also responsible for defining the type of objects to be parsed.

### Extended Backus-Naur Form Expressions
Parser rules have access to a more restricted set of EBNF expressions compared to terminal rules. Parser and terminal rules share the following expressions:
- Groups
- Alternatives

> - Keywords

Reviewer 2: Terminal rules are also able to use keywords. `terminal X: 'Y';` is a valid terminal.

Reviewer: Right, but we don't call those keywords. They are merely character sequences (like groups): `'xyz'` is the same as `'x' 'y' 'z'`.

        'person' name=ID
    ```
    Here, the property `name` will accept only one expression matching the terminal rule `ID`.

2. `+=` is used to assign **multiple expressions** to a list property
    ```
    Paragraph:
        'paragraph' (sentences+=STRING)+;
    ```
    Here, the property `sentences` will accept one or many expressions matching the terminal rule `STRING`. The expressions belong to a single group and must not be interrupted by other expressions. The parsing of a document containing:
    ```
    paragraph "The expression group " 3 "was interrupted"
    ```
    will throw an error since the `STRING` expressions are not continuous. It is however possible to interrupt and resume a sequence of expressions by using [hidden terminal symbols](#hidden-terminal-symbols):
    ```
    paragraph "expression one" /* comment */ "expression two"
    ```
    The above example will be successfully parsed.

3. `?=` is used to assign a **value to a property of type boolean**. The value of the property of type `boolean` is set to `true` if the right part of the assignment is consumed by the parser.
    ```
    Employee:
        'employee' name=ID (remote?='remote')?
    ```

>     Here the value of the property `remote` will be set to true if the keyword `remote` appears in the document.

Reviewer 2: Minor: This is a bit misleading.

```suggestion
    Here the value of the property `remote` will be set to `true` if the keyword `remote` was successfully parsed as a part of the rule call.
```

#### Cross-reference

> With Langium, you can declare *cross-references* directly in the grammar. *Cross-reference* allows to reference an object of a given type. The syntax is:

Reviewer 2: Minor:

```suggestion
With Langium, you can declare *cross-references* directly in the grammar. A *cross-reference* allows to reference an object of a given type. The syntax is:
```

```
property=[Type:TOKEN]
```

> The `property` will be a reference to an object of type `Type` identified by the token `TOKEN`. If the `TOKEN` is omitted, the parser assumes that it is matching the terminal rule `ID`.

Reviewer 2: Since we use a lot of functionality in this documentation which is only available in version 0.3, we should use the 0.3 behavior here as well:

```suggestion
The `property` will be a reference to an object of type `Type` identified by the token `TOKEN`. If the `TOKEN` is omitted, the parser will use the terminal or data type rule associated with the `name` assignment of the `Type` rule.
```

```
Person:
    'person' name=ID;
Greeting:
    'Hello' person=[Person:ID] '!';
```
The `Person` in square brackets does not refer to the parser rule `Person` but instead refers to an object of type `Person`. It will successfully parse a document like:
```
person Bob
Hello Bob !
```

> but will fail to parse:

Reviewer 2: Suggestion: The _parse_ will not fail, since this isn't a parsing error. Instead, the cross reference resolution will fail, resulting an error message.

Author: https://github.com/langium/langium-website/pull/21#:~:text=cross%20reference%20resolution-,847ceb2,-Merge%20state

```
person Bob
Hello Sara !
```
because a `Person` object with the ID of 'Sara' has not been instantiated even though 'Sara' is a valid `ID`.

#### Unordered Groups
[NOTE] Unordered group are currently not supported but seems like it's being worked on. I think this is still valuable info that could be added to the documentation at a later stage if needed.

By default, a parser rule has to be implemented in the exact order it is declared.
```
Person:
    'person' name=ID age=INT
```
Here a `Person` object **needs** to first declare the property `name` then `age`.
```
person Bob 25
```
will successfully be parsed to an object of type `Person` while
```
person 25 Bob
```
will throw an error.

However, it is possible to declare a group of properties in an unordered fashion using the `&` operator
```
Person:
    'person' name=ID & age=INT
```
will now allow `name` and `age` to be declared in any order.
```
person 25 Bob
```
will then successfully create an object of type `Person`.

> Cardinality (?,*,+ operators) also applies to unordered group. Please note that assignments with a cardinality of `+` or `*` have to appear continuously and cannot be interrupted by an other assignment and resumed later.

Reviewer 2: Minor: Escape usage of `*` using `\*`. This breaks markdown formatting.

Author: It looks fine on the Netlify preview

    'keywordOne' name=ID | RuleTwo;

RuleTwo returns TypeTwo:
    'keywordTwo' name=ID;
```
In the above example, we rely on a *rule call* to specify the return type. With more complex rules, the readability will be highly impacted. *Actions* allow to improve the readability of the grammar by explicitly defining the return type. *Actions* are declared inside of curly brackets `{}`:
```
RuleName returns TypeOne:
    'keywordOne' name=ID | 'keywordTwo' {TypeTwo} name=ID;
```

#### Unassigned Rule Calls
Parser rules do not necessarily need to return an object, they can also refer to other parser rules which in turn will be responsible for returning the object.
For example, in the [Arithmetics example](https://github.com/langium/langium/blob/main/examples/arithmetics/src/language-server/arithmetics.langium):
```
AbstractDefinition:
	Definition | DeclaredParameter;
```
The parser rule `AbstractDefinition` will not create an object of type AbstractDefinition. Instead, it calls either the `Definition` or `DeclaredParameter` parser rule which will be responsible for creating an object of a given type (or call other parser rules if they are unassigned rule calls themselves). 

> #### Assigned Actions

Reviewer 2: ```suggestion
#### Assignments
```

Author: There is already a section "Assignments" (line ~190) before where it is described how to assign data to a property/feature. 
Any idea on renaming that wouldn't be confusing with the previous section?

The parser is built using [Chevrotain](https://github.com/chevrotain/chevrotain) which implements a LL(k) parsing algorithm (left-to-right). By definition, a LL(k) grammar cannot have rules containing left recursion.

Consider the following: 
```
Addition:
    Addition '+' Addition | '(' Addition ')' | value=INT;
```
The parser rule `Addition` is left-recursive and will not be parsable. We can go around this issue by *left-factoring* the rule, *i.e.* by factoring out the common left-factor. We introduce a new rule `Expression`:
```
Expression:
    '(' Addition ')' | value=INT;

Addition:
    Expression ('+' Expression)*;
```

[NOTE] I left out the part regarding `tree rewrite actions` as I think it could be part of a more advanced documentation part on the AST/CST. I can add it if we think it belongs here.

> ### Syntactic Predicates

Reviewer 2: Syntactic predicates were only needed in Xtext due to limitations of the underlying parser generator (Antlr3). We don't need syntactic predicates in Langium. We support them, but only for backwards compatibility.

Author: Should we then mention that it is supported for backwards compatibility but not recommend it for new grammars? Can we then just link to the Xtext documentation?

Reviewer 2: The grammar is backwards compatible in a way that enables syntactic predicates from a grammar parser perspective , but it does not do anything in any way. So I think we shouldn't mention it at all.

Author: Removed

The following example from the [domain model example]() avoids for the `QualifiedName` data type rule to conflict with the terminal rule `ID`.
```
QualifiedName returns string:
    ID ('.' ID)*;
```
Data type rules need to specify a primitive return type.

## The Entry Rule
The *entry rule* is a parser rule that defines the starting point of the parsing step. The *entry rule* starts with the keyword `entry` and matches other parser rules.
 ```
 entry Model:
    (persons+=Person | greetings+=Greeting)*;
 ```
In this example, the *entry rule* `Model` defines a group of alternatives. The parser will go through the input document and try to parse a `Person` or a `Greeting` object and add it to the array `persons` or `greetings`, respectively. Thanks to the grouping and the cardinality of zero or many (`*`), the parser will go through the document until all inputs have been consumed.

> Not all parser rules need to be registered in the parser rule.

Reviewer 2: Minor:

```suggestion
Not all parser rules need to be registered in the entry parser rule.
```
